lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,291 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from base64 import b64encode
+from collections import namedtuple
+from hashlib import sha256
+import json
+import logging
+from operator import attrgetter
+
+from django.db import models
+from django.db.utils import IntegrityError
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+
+log = logging.getLogger(__name__)
+
+
+# Used to serialize information about a block at the time it was used in
+# grade calculation.
+BlockRecord = namedtuple('BlockRecord', ['locator', 'weight', 'max_score'])
+
+
+class BlockRecordSet(frozenset):
+    """
+    An immutable ordered collection of BlockRecord objects.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(BlockRecordSet, self).__init__(*args, **kwargs)
+        self._json = None
+        self._hash = None
+
+    def to_json(self):
+        """
+        Return a JSON-serialized version of the list of block records, using a
+        stable ordering.
+        """
+        if self._json is None:
+            sorted_blocks = sorted(self, key=attrgetter('locator'))
+            # Remove spaces from separators for more compact representation
+            self._json = json.dumps(
+                [block._asdict() for block in sorted_blocks],
+                separators=(',', ':'),
+                sort_keys=True,
+            )
+        return self._json
+
+    @classmethod
+    def from_json(cls, blockrecord_json):
+        """
+        Return a BlockRecordSet from a json list.
+        """
+        block_dicts = json.loads(blockrecord_json)
+        record_generator = (BlockRecord(**block) for block in block_dicts)
+        return cls(record_generator)
+
+    def to_hash(self):
+        """
+        Return a hashed version of the list of block records.
+
+        This currently hashes using sha256, and returns a base64 encoded version
+        of the binary digest.  In the future, different algorithms could be
+        supported by adding a label indicated which algorithm was used, e.g.,
+        "sha1$witfkXg0JglCjW9RssWvTAveakI=".
+        """
+        if self._hash is None:
+            self._hash = b64encode(sha256(self.to_json()).digest())
```

Comment:
Sorry for the nit, but please use SHA1, especially since it's going to be in every fkey from the grade storage table.